### ISSUE TITLE
Proper way to detect if eclipse workspace is locked

### ISSUE THREAD

**Issue by hohwille** [devonfw/devon-ide#41]
We need to figure out if an eclipse workspace is locked or not
* https://bugs.eclipse.org/bugs/show_bug.cgi?id=197428
* https://stackoverflow.com/questions/33431762/how-to-check-if-eclipse-workspace-is-locked/33432112?noredirect=1#comment96829310_33432112
* https://stackoverflow.com/questions/14266016/why-doesnt-filechannel-trylock-prevent-access-across-multiple-processes

**Comment by markusschuh** [comment]
First challenge: eclipse supports [different lock mechanisms:](https://help.eclipse.org/2019-03/topic/org.eclipse.platform.doc.isv/reference/misc/runtime-options.html?#osgilocking)  
The default -  `-Dosgi.locking=java.nio` seems to be then one, used by devon. (checked on Windows 7). But this may be also "java.io" on macOS / linux.

The good part: If java.nio file locking is used, on windows this uses a file handle, also visible by windows. And Users have the right to read out their filehandles. But that needs a tool to access the kernel - even WMI has no call for the file handles. (what would allow some command line use)

I have tried the check with help of a some java code, using tryLock() of [java.nio.channels.FileLock](https://docs.oracle.com/javase/8/docs/api/java/nio/channels/AsynchronousFileChannel.html#tryLock-long-long-boolean-) but this works.

It would work with external tool "handle.exe" from [sysinternals](https://live.sysinternals.com) - owned by Microsoft. But even then it is nasty to script that with help of windows CMD:

    :: untested code
    set path_lockfile=%WORKSPACE_PATH%\.metadata\.lock
    if exist %path_lockfile% (
      :: findstr needs double backslashes
      set path_lockfile_for_search=%path_lockfile:\=\\%
      for /f "usebackq tokens=1-3,*" %%P in (`handle64 -nobanner -p javaw.exe ^| "%SystemRoot%\system32\bin\findstr" /R "^javaw.exe %path_lockfile_for_search%$"`) do (
        :: not yet realized: handle output like below. 
        :: javaw.exe pid:  <thispid> <domain\username>
        ::  <lockhandle>: File          <filename>
        :: javaw.exe pid:  <anotherjavaw_pid> <domain\username>
        ::javaw.exe pid:  <anotherjavaw_pid> <domain\username>
      )
    )

Meta Code:
Loop over results:
   for lines with token1 == javaw.exe set var javaw_pid
   for lines with token3 == %path_lockfile_for_search% set var "eclipse_running_owning_lock"= value of javaw_pid and break loop

If there is a file - but no matching eclipse there should not be a lock and file can be removed.
If there is a file and a javaw.exe with a pid owning the lock, there is a running eclipse, which locks that workspace. If there is no eclipse GUI visible for the user (this could happen in some error cases, it might help to launch some wm message to the javaw.exe - or its parent process eclipse.exe?  

Those tasks are even two more challenges that would need to resvoled.

**Comment by hohwille** [comment]
This eclipse locking mechanism is kind of strange:
1. when eclipse shuts down regularly it does not remove the `.lock` file
2. while eclipse is running and should hold the lock on that `.lock` file, I can easily remove it (on mac  OS with `rm .metadata/.lock`). If I then launch eclipse I have two instances running on the same workspace in parallel what is actually what the locking should prevent.

**Comment by hohwille** [comment]
I can try to include the Java Code from eclipse to do the lock. If I get the exception that eclipse is already locked, then I properly detected that the workspace is locked. However, otherwise I successfully locked the workspace and have to unlock it again as I actually did not wanted to lock it but just wanted to check if it is locked. In some edge-cases this could cause other problems...
However, this would at least be an OS independent solution what is IMHO better than create one for every OS what might break with new Eclipse versions and will be hard to maintain and test.
